gobblin-runtime/src/main/java/org/apache/gobblin/quality/DataQualityEvaluator.java

```diff
@@ -0,0 +1,221 @@
+        DataQualityStatus jobDataQualityStatus = DataQualityStatus.PASSED;
+        int totalFiles = 0;
+        int failedFilesCount = 0;
+        int passedFilesCount = 0;
+        int nonEvaluatedFilesCount = 0;
+
+        for (TaskState taskState : taskStates) {
+            totalFiles++;
+
+            // Handle null task states gracefully
+            if (taskState == null) {
+                log.warn("Encountered null task state, skipping data quality evaluation for this task");
+                nonEvaluatedFilesCount++;
+                continue;
+            }
+
+            DataQualityStatus taskDataQuality = null;
+            String result = taskState.getProp(ConfigurationKeys.TASK_LEVEL_POLICY_RESULT_KEY);
+            taskDataQuality = DataQualityStatus.fromString(result);
+            if (taskDataQuality != DataQualityStatus.NOT_EVALUATED) {
+                log.debug("Data quality status of this task is: " + taskDataQuality);
+                if (DataQualityStatus.PASSED == taskDataQuality) {
+                    passedFilesCount++;
+                } else if (DataQualityStatus.FAILED == taskDataQuality){
+                    failedFilesCount++;
+                    jobDataQualityStatus = DataQualityStatus.FAILED;
+                } else {
+                    log.warn("Unexpected data quality status: " + taskDataQuality + " for task: " + taskState.getTaskId());
+                }
+            } else {
+                // Handle files without data quality evaluation
+                nonEvaluatedFilesCount++;
+                log.warn("No data quality evaluation for task: " + taskState.getTaskId());
+            }
+        }
+
+        // Log summary of evaluation
+        log.info("Data quality evaluation summary - Total: {}, Passed: {}, Failed: {}, Not Evaluated: {}",
+            totalFiles, passedFilesCount, failedFilesCount, nonEvaluatedFilesCount);
+        return new DataQualityEvaluationResult(jobDataQualityStatus, totalFiles, passedFilesCount, failedFilesCount, nonEvaluatedFilesCount);
+    }
+
+
+    private static void emitMetrics(JobState jobState, final DataQualityStatus jobDataQuality, final int totalFiles,
+            final int passedFilesCount, final int failedFilesCount, final int nonEvaluatedFilesCount, final String datasetUrn) {
+        try {
+            // Check if OpenTelemetry is enabled
+            boolean otelEnabled = jobState.getPropAsBoolean(ConfigurationKeys.METRICS_REPORTING_OPENTELEMETRY_ENABLED,
+                ConfigurationKeys.DEFAULT_METRICS_REPORTING_OPENTELEMETRY_ENABLED);
+
+            if (!otelEnabled) {
+                log.info("OpenTelemetry metrics disabled, skipping metrics emission");
+                return;
+            }
+
+            OpenTelemetryMetricsBase otelMetrics = OpenTelemetryMetrics.getInstance(jobState);
+            log.info("OpenTelemetry instance obtained: {}", otelMetrics != null);
+
+            if (otelMetrics != null) {
```

Comment:
we can check for `if (otelMetrics == null)` and return early. It also simplifies the rest of the method by reducing indentation and nesting, making the main logic easier to follow.